text/0000-otlp-proto-version-identifier.md

```diff
@@ -0,0 +1,33 @@
+# Include OTLP proto version identifier in requests
+
+Add OTLP request identifier to describe the OTLP version used to create the binary or JSON payload.
+
+## Motivation
+
+It is useful to know the version of the OTLP protobuf definition used to create the binary or JSON, without having to decode the payload.
+
+This would enable routing to different OTLP handlers by inspecting the identifier value. This is especially useful for organisations that wish to support multiple OTLP versions where the proto definitions have changed. A currently used alternative is to append the version to the path, but this is clumsy and requires additional configuration when setting up the exporter.
+
+## Explanation
+
+Currently it is not possible to know the version of the OTLP proto definition before receiving it. This is challenging because there have been changes to the OTLP proto specification that are breaking.
```

Comment:
I think this phrasing indirectly normalizes the fact that there may be breaking changes, which may contribute to our sloppiness and indirectly cause breakage. I think we should do the opposite and emphasize even stronger that breaking changes are extremely bad and we should never do them without officially introducing a new major OTLP version (in which case the version number outside the payload may be introduced).

My opinion is still that this OTEP and the version number outside the payload moves us in the wrong direction.
Instead I think we should make a bigger effort to avoid breaking changes, for example by adding tooling that catches it, by making it clearer in our spec/docs what changes constitute breaking changes and should avoided, etc.

Reply:
I understand and fully appreciate the intention to avoid breaking changes to the wire format. However, we've just seen another _breaking_ change be merged into the [OTLP spec](https://github.com/open-telemetry/opentelemetry-proto/pull/407).

Removing deprecated fields like these forces OTLP receivers to chose between either supporting senders who have not upgraded to a newer version of the OTLP spec (<0.12) or supporting senders who have upgraded(0.19+). Not everyone has the luxury of dictating when senders upgrade. These deprecated fields could have existed in the spec until a major release (eg 1.0) and receivers can choose to support them if they wish to.

A solution to the above is for receivers to utilise multiple handlers using different OTLP formats. However, as we've discussed previous, it's cumbersome and inefficient to decode a request payload to get the version to then pass onto one of these handlers which will likely decode again. An identifier outside of the payload allows receivers to decide how they handle the request before using resources on decoding.

Reply:
I see a "Why not both?" solution. 

We should certainly strive for all the bigger, ideal efforts @tigrannajaryan describes: avoid proto breakage, more tooling to detect breakage. If we also include the proto version in the request and outside of the body (an `x-otlp-version` header? `otlp/1.2.3` included in [exporter user_agent](https://github.com/open-telemetry/opentelemetry-specification/pull/2684)?), receivers of OTLP have easy access to the data necessary to troubleshoot communications problems when we haven't lived up to that ideal.